core/src/main/java/com/yahoo/oak/Chunk.java

```diff
@@ -788,25 +860,62 @@ public boolean hasNext() {
         @Override
         public int next(ThreadContext ctx) {
             int toReturn = next;
-            advance();
+            advance(ctx);
             return toReturn;
         }
-    }
 
-    class DescendingIter implements ChunkIter {
+        @Override
+        public boolean isBoundCheckNeeded() {
+            return needBoundCheckDynamic;
+        }
 
-        private int next;
+        private int advanceNextIndex(final int entryIndex, ThreadContext ctx) {
+            int next = entryIndex;
+            while (next != NONE_NEXT && !entrySet.isValueRefValidAndNotDeleted(next)) {
+                next = entrySet.getNextEntryIndex(next);
+                if (needBoundCheckStatic && !needBoundCheckDynamic && next==midIdx) {
+                    // needBoundCheckDynamic false, but needBoundCheckStatic is true
+                    // need to reset needBoundCheckDynamic to true when reaching the midIndex
+                    needBoundCheckDynamic = true;
+                }
+            }
+            return next;
+        }
+
+        private int advanceNextIndexNoBound(final int entryIndex, ThreadContext ctx) {
+            int next = entryIndex;
+            while (next != NONE_NEXT && !entrySet.isValueRefValidAndNotDeleted(next)) {
+                next = entrySet.getNextEntryIndex(next);
+            }
+            return next;
+        }
+
+        @Override
+        protected boolean afterStop(OakScopedReadBuffer key) {
+            if (to == null) {
+                return false;
+            }
+            if (key == null) {
+                // we are on the last chunk and 'to' is not null
+                return true;
+            }
+            int c = comparator.compareKeyAndSerializedKey(to, key);
+            // return true if to<key or to==key and the scan was not fromInclusive
+            return c < 0 || (c == 0 && !toInclusive);
+        }
+    }
+
+    class DescendingIter extends ChunkIter {
         private int anchor;
         private int prevAnchor;
         private final IntStack stack;
         private final K from;
-        private boolean inclusive;
+        private boolean fromInclusive;
+        private final int skipEntriesForBiggerStack = (maxItems/10); // 1 is the lowest possible value
```

Comment:
Correctness: if `maxItems` is less than 10, `skipEntriesForBiggerStack` will be zero.

```suggestion
        private final int skipEntriesForBiggerStack = Math.max(1, maxItems / 10); // 1 is the lowest possible value
```